### ISSUE TITLE
Doc : Access to instance positions in NodeMaterial

### ISSUE THREAD

**Issue by Makio64** [mrdoob/three.js#29071]
### Description

I coudnt find the right way to access or use instance properties in TSL from the doc (  [Three.js-Shading-Language](https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language) ) or examples.

Would be nice to get the explanation on how to access to the `ID` and `instanceMatrix` / `instanceColorMatrix` and i would be happy to create simple examples and update the documentation.

### Solution

- Update the doc for TSL including accessors for instance / batch
- Add minimal examples : The color of the material change depending of the x y z axis of the instance

### Alternatives

Another minimal example : the scale of the instance change depending of the luminosity of the instanceColorMatrix

**Comment by cmhhelgeson** [comment]
Hi @Makio64. If you want to access the instance of a mesh within a TSL function, you need to use instanceIndex. Now within the examples, instanceIndex is largely used within compute shaders but the value it represents changes depending on what context it is used in. Within a compute shader ( called via a .compute() call ), instanceIndex represents the invocation id ( or execution thread ) that the compute shader is currently running. However, within a TSL function that is being used as a vertex shader ( i.e one  that is applied to a material's positionNode or vertexNode, instanceIndex will represent the instance of the geometry the vertex belongs to.

Here's a quick example I wrote up demonstrating how to use instanceIndex within a TSL vertex shader.

```
	const material = new THREE.MeshBasicNodeMaterial( {map: texture});


	const uCircleRadius = uniform(1.0);
	const uCircleSpeed = uniform( 0.5 );
	const instanceCount = 80;
	const numCircles = 4;
	const meshesPerCircle = instanceCount / numCircles;

	material.positionNode = tslFn(() => {
                // Multiple elapsed time by speed
		const time = timerLocal().mul( uCircleSpeed );

		const instanceWithinCircle = instanceIndex.remainder(meshesPerCircle );

		const circleIndex = instanceIndex.div(meshesPerCircle ).add(1);

		const circleRadius = uCircleRadius.mul(circleIndex);

		// Normalize instanceIndex to range [0, 2*PI]
		const angle = float(instanceWithinCircle).div(meshesPerCirlce).mul(PI2).add( time ); 
		const circleX = sin( angle ).mul( circleRadius );
		const circleY = cos( angle ).mul( circleRadius );

		const scalePosition = positionGeometry.mul( circleIndex );
		
		const rotatePosition = rotate(scalePosition, vec3( timerLocal() , timerLocal(), 0.0));

		const newPosition = rotatePosition.add( vec3( circleX, circleY, 0.0 ));
		return vec4( newPosition, 1.0 );
	})(); 

	mesh = new THREE.InstancedMesh( geometry, material, instanceCount );
	console.log(mesh)
	scene.add( mesh );

	renderer = new THREE.WebGPURenderer({ antialias: false })
	renderer.setPixelRatio( window.devicePixelRatio );
	renderer.setSize( window.innerWidth, window.innerHeight );
	renderer.setAnimationLoop( animate );
	document.body.appendChild( renderer.domElement );
	//
	
	const controls = new OrbitControls( camera, renderer.domElement );
	controls.minDistance = 1;
	controls.maxDistance = 20;

	const gui = new GUI();
	gui.add( uCircleRadius, 'value', 0.1, 3.0, 0.1 ).name( 'Circle Radius' );
	gui.add( uCircleSpeed, 'value', 0.1, 3.0, 0.1 ).name( 'Circle Speed' );
```

And the output:

**Comment by Makio64** [comment]
Thanks @cmhhelgeson, nice examples, very instructive !

In my case I still need to access to the : `instanceMatrix` to know the position of the instance in the worldSpace ( basically to apply position modifier on the vertex depending of the position of the vertex in worldspace )  

I tried to access it using `{instance} from 'three/tsl'` but there is a logic I cant figure out.

The easy solution would be to do a `InstancedBufferAttribute` or `StorageInstancedBufferAttribute` with the data I need ( position of the instance ) but as we already have an instanceMatrix system so I would like to understand how to access and use it.

**Comment by cmhhelgeson** [comment]
> Thanks @cmhhelgeson, nice examples, very instructive !
> 
> In my case I still need to access to the : `instanceMatrix` to know the position of the instance in the worldSpace ( basically to apply position modifier on the vertex depending of the position of the vertex in worldspace )
> 
> I tried to access it using `{instance} from 'three/tsl'` but there is a logic I cant figure out.
> 
> The easy solution would be to do a `InstancedBufferAttribute` or `StorageInstancedBufferAttribute` with the data I need ( position of the instance ) but as we already have an instanceMatrix system so I would like to understand how to access and use it.

I myself wasn't even aware there was an instance node 😅.

I'm not quite sure what the requirements of your application are, but I think a good place to start would be using the positionWorld node (I can get back on whether the modelWorldMatrix used in positionWorld updates to account for instancing). As you mentioned, you could also just replace the current InstanceMatrix InstancedBufferAttribute with a StorageInstancedBufferAttribute, thus making it accessible within TSL shaders. Look at webgpu_compute_geometry.html to see how StorageBufferAttributes can easily replace existing buffer attributes.

**Comment by Makio64** [comment]
@z4122 @sunag do you mind sharing an example of how to use the InstanceNode / access to the instance matrix from the tsl ? 🥺 

Thanks !

**Comment by sunag** [comment]
Currently `InstanceNode` does not have feature to just export `instanceMatrixNode` property, maybe you will have to open the source code and use the separate code part for now.

```js
import { buffer, instanceIndex } from 'three/tsl';

// if ( instanceMesh.count <= 1000 )
const node = buffer( instanceMesh.instanceMatrix.array, 'mat4', instanceMesh.count ).element( instanceIndex );
```

https://github.com/mrdoob/three.js/blob/410f73742f0043540857a60d3827e4635875ec90/src/nodes/accessors/InstanceNode.js#L41-L73

**Comment by Mugen87** [comment]
Documentation related issues in context of TSL can be merged into #29829.
